share/namespace.go

```diff
@@ -0,0 +1,134 @@
+package share
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	appns "github.com/celestiaorg/celestia-app/pkg/namespace"
+	"github.com/celestiaorg/nmt/namespace"
+)
+
+// Various reserved namespaces.
+var (
+	MaxReservedNamespace     = appns.MaxReservedNamespace.Bytes()
+	ParitySharesNamespace    = appns.ParitySharesNamespace.Bytes()
+	TailPaddingNamespace     = appns.TailPaddingNamespace.Bytes()
+	ReservedPaddingNamespace = appns.ReservedPaddingNamespace.Bytes()
+	TxNamespace              = appns.TxNamespace.Bytes()
+	PayForBlobNamespace      = appns.PayForBlobNamespace.Bytes()
+)
+
+// Namespace represents namespace of a Share.
+// Consists of version byte and namespace ID.
+type Namespace []byte
+
+// NamespaceFromBytes converts bytes into Namespace and validates it.
+func NamespaceFromBytes(b []byte) (Namespace, error) {
+	n := Namespace(b)
+	return n, n.Validate()
+}
+
+// Version reports version of the Namespace.
+func (n Namespace) Version() byte {
+	return n[appns.NamespaceVersionSize-1]
+}
+
+// ID reports version of the Namespace.
+func (n Namespace) ID() namespace.ID {
+	return namespace.ID(n[appns.NamespaceVersionSize:])
+}
+
+// String stringifies the Namespace.
+func (n Namespace) String() string {
+	return hex.EncodeToString(n)
+}
+
+// Validate checks if the namespace is correct.
+func (n Namespace) Validate() error {
+	if len(n) != NamespaceSize {
+		return fmt.Errorf("invalid namespace length: %v must be %v", len(n), NamespaceSize)
+	}
+	if n.Version() != appns.NamespaceVersionZero && n.Version() != appns.NamespaceVersionMax {
+		return fmt.Errorf("unsupported namespace version %v", n.Version())
+	}
+	if len(n.ID()) != appns.NamespaceIDSize {
+		return fmt.Errorf("unsupported namespace id length: id %v must be %v bytes but it was %v bytes",
+			n.ID(), appns.NamespaceIDSize, len(n.ID()))
+	}
+	if n.Version() == appns.NamespaceVersionZero && !bytes.HasPrefix(n.ID(), appns.NamespaceVersionZeroPrefix) {
+		return fmt.Errorf("unsupported namespace id with version %v. GetNamespace %v must start with %v leading zeros",
+			n.Version(), n.ID(), len(appns.NamespaceVersionZeroPrefix))
+	}
+	return nil
+}
+
+// ValidateBlobNamespace returns an error if this namespace is not a valid blob namespace.
+func (n Namespace) ValidateBlobNamespace() error {
+	if err := n.Validate(); err != nil {
+		return err
+	}
+	if n.IsReserved() {
+		return fmt.Errorf("invalid blob namespace: %v cannot use a reserved namespace, want > %v",
+			n, appns.MaxReservedNamespace.Bytes())
+	}
+	if n.IsParityShares() {
+		return fmt.Errorf("invalid blob namespace: %v cannot use parity shares namespace", n)
+	}
+	if n.IsTailPadding() {
+		return fmt.Errorf("invalid blob namespace: %v cannot use tail padding namespace", n)
+	}
+	return nil
+}
+
+func (n Namespace) IsReserved() bool {
+	return bytes.Compare(n, MaxReservedNamespace) < 1
```

Comment:
It works over hashes, not a namespace.

Reply:
Also, I moved these helpers as methods. Much much nicer now

Reply:
i was referring to namespace methods:
```
func (n Namespace) IsLessThan(n2 Namespace)
```

Also you would need to make MaxReservedNamespace, ParitySharesNamespace, etc a namespace type. I think it would be nicer for them to be Namespace type too, so user will be able to use it like:
```
mynaspace. Equals(ParitySharesNamespace)
```
And there will be no need for explicit method for every namespace

Reply:
Yes, but it is a type of `[]byte` not `Namespace`. I suggest to make it latter, so it is possible to use all new introduced tooling associated with type.

Reply:
> func (n Namespace) IsLessThan(n2 Namespace)

I tried this in the integration and didn't really like it. It's much cleaner on the user side to take the whole hash. You will see

Reply:
> And there will be no need for explicit method for every namespace

I like this idea though.

Reply:
> Yes, but it is a type of []byte not Namespace. I suggest to make it latter, so it is possible to use all new introduced tooling associated with type.

Golang takes care of this automatically. You don't need to define. I think this is the case since generic

Reply:
I am using Equals now on these vars that are []byte

Reply:
I am not sure we need this, but It does not hurt to have it.

Reply:
I think we should do it just to be consistent with types